Import important libraries

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd 

# 2. Exploring data (often interleaved with preprocessing)

Read all .csv files and perform exploration.

We decide that we wouldn't use the book_id.csv file since it is redundant

First, we will read data file "book_data.csv" and save to DataFrame `book_df`. 

In [2]:
book_df = pd.read_csv('data/book_data.csv')
book_df.head()

product_id                            title                    authors  \
0    74021317             Cây Cam Ngọt Của Tôi  José Mauro de Vasconcelos   
1    74021317             Cây Cam Ngọt Của Tôi  José Mauro de Vasconcelos   
2   184466860  Hành Tinh Của Một Kẻ Nghĩ Nhiều       Nguyễn Đoàn Minh Thư   
3    73787185         Những Tù Nhân Của Địa Lý               Tim Marshall   
4    52789367       Nhà Giả Kim (Tái Bản 2020)               Paulo Coelho   

   original_price  current_price  quantity                    category  \
0          108000          64800   53075.0                 Tiểu Thuyết   
1          108000          64800   53075.0                 Tiểu Thuyết   
2           86000          59900    7929.0  Sách tư duy - Kỹ năng sống   
3          210000         126000   17896.0               Lĩnh vực khác   
4           79000          47400   24668.0          Tác phẩm kinh điển   

   n_review  avg_rating pages              manufacturer  \
0     11481         5.0   244  Nhà Xuất Bản Hội Nhà Văn   
1     11481         5.0   244  Nhà Xuất Bản Hội Nhà Văn   
2       780         4.8   184     Nhà Xuất Bản Thế Giới   
3      3623         4.8   430  Nhà Xuất Bản Hội Nhà Văn   
4      5131         5.0   228       Nhà Xuất Bản Hà Nội   

                                          cover_link  
0  https://salt.tikicdn.com/ts/product/5e/18/24/2...  
1  https://salt.tikicdn.com/ts/product/5e/18/24/2...  
2  https://salt.tikicdn.com/ts/product/d7/99/24/1...  
3  https://salt.tikicdn.com/ts/product/8d/96/9e/c...  
4  https://salt.tikicdn.com/ts/product/45/3b/fc/a...

Second, we will read data file "comment_data.csv" and save to DataFrame `comment_df`. 

In [3]:
comment_df = pd.read_csv('data/comments.csv')
comment_df.head()

product_id  comment_id            title  thank_count  customer_id  rating  \
0    74021317    12559756  Cực kì hài lòng          313     22051463       5   
1    74021317    16979365  Cực kì hài lòng            6     27791831       5   
2    74021317    14069617  Cực kì hài lòng           25     17748750       5   
3    74021317     8569824  Cực kì hài lòng           57       410797       5   
4    74021317    18368714  Cực kì hài lòng            0     28545286       5   

                                             content  
0  Có những người bước đến, họ lấp đầy hạnh phúc ...  
1  Thấy nhiều bạn chê tiki gói hàng quá, may sao ...  
2  Bìa cực xinh, tiki giao hàng nhanh, sách không...  
3  Sách lúc nhận hình thức rất ổn, không cong vên...  
4  Một cuốn sách rất đáng đọc về tình yêu thương ...

### 2.1 How many rows and how many columns?

We get the number of rows and columns of the DataFrame `book_df` and store it in `num_rows` and `num_cols` variables.

In [4]:
num_rows = book_df.shape[0]
num_cols = book_df.shape[1]
print(f'Number of rows: ',num_rows)
print(f'Number of cols: ',num_cols)

Number of rows:  1796
Number of cols:  12


We get the number of rows and columns of the DataFrame `comment_df` and store it in `num_rows` and `num_cols` variables.

In [5]:
num_rows = comment_df.shape[0]
num_cols = comment_df.shape[1]
print(f'Number of rows: ',num_rows)
print(f'Number of cols: ',num_cols)

Number of rows:  141281
Number of cols:  7


### 2.2 What is the meaning of each column?

Meaning of each column in `book_df` describle below

| Column      | Description |
| :---        |    :----   |    
| `product_id`      | Id of the product in the Tiki database (unique)      | 
| `title`   |  Name of the book, maybe contain republish time       | 
| `authors`    |    same with it's name              |
| `original_price`    |      price at the first time             |
| `current_price`    |  price at present if having a discount                 |
| `quantity`    |  	total number of books sold of all time                 |
| `category`    |  	kind of book                 |
| `n_review`    |   number of reviews               |
| `avg_rating`    |  average rating (max 5.0)                 |
| `pages`    |     total pages of each book              |


Meaning of each column in `comment_df` describle below

| Column      | Description |
| :---        |    :----   |    
| `product_id`      | same with book_data file      | 
| `title`   |  keywork of comment       | 
| `comment_id`    |   each comment has individual id        |
| `thank_count`    |     number of like of other people            |
| `customer_id`    |  price at present if having a discount                 |
| `rating`    |  average rating of the comment             |
| `category`    |  	kind of book                 |
| `content`    |  same with it's name             |


### 2.3 Are there duplicated rows?

In [6]:
num_duplicated_rows = book_df[book_df.duplicated(keep=False)].shape[0]
num_duplicated_rows

35

In `book_df` there are 35 duplicated rows. So we will remove it

In [7]:
book_df = book_df.drop_duplicates()
num_rows = book_df.shape[0]
num_rows

1777

After remove duplicated rows, `book_df` has 1777 rows

In [8]:
num_duplicated_rows = comment_df[comment_df.duplicated(keep=False)].shape[0]
num_duplicated_rows

39299

In `comment_df` there are 39299 duplicated rows. So we will remove it

In [9]:
comment_df = comment_df.drop_duplicates()
num_rows = comment_df.shape[0]
num_rows

121393

After remove duplicated rows, `comment_df` has 121393 rows

### 2.4 What is the current data type of each column? Are there columns having inappropriate data types?

In [10]:
dtypes = pd.Series(dict(book_df.dtypes))
dtypes

product_id          int64
title              object
authors            object
original_price      int64
current_price       int64
quantity          float64
category           object
n_review            int64
avg_rating        float64
pages              object
manufacturer       object
cover_link         object
dtype: object

We can see some of data type of the column has appropriate type. Some of them in object type like `pages`,`manufacturer`, `category`, `title`, `authors`,... We will explore and convert pages into integer and the remain will be string

In [11]:
# book_df.astype({'pages': np.float32})
book_df['pages'] = pd.to_numeric(book_df['pages'] , errors='coerce')

In [12]:
def open_object_dtype(s):
    dtypes = set()
    s.apply(lambda x: dtypes.add(type(x)))
    return dtypes
print('category',open_object_dtype(book_df.category) )
print('manufacturer',open_object_dtype(book_df.manufacturer) )
book_df.loc[book_df['category'].str.strip() == '','category'] = np.nan
book_df.loc[book_df['manufacturer'].str.strip() == '','manufacturer'] = np.nan
book_df = book_df.astype({'manufacturer': str,'category':str})

category {<class 'str'>}
manufacturer {<class 'str'>, <class 'float'>}


We will replace which rows has zero characters by `np.nan` and cast them into str

In [13]:
dtypes = pd.Series(dict(comment_df.dtypes))
dtypes

product_id      int64
comment_id      int64
title          object
thank_count     int64
customer_id     int64
rating          int64
content        object
dtype: object

With `comment_df` we also see some of data type of the column has appropriate type. Some of them in object type like `title`,`content`. We will explore and convert them into string

In [14]:
print('title',open_object_dtype(comment_df.title ) )
print('content',open_object_dtype(comment_df.content ) )
comment_df.loc[comment_df['title'].str.strip() == '','title'] = np.nan
comment_df.loc[comment_df['content'].str.strip() == '','content'] = np.nan
comment_df = comment_df.astype({'title': str,'content':str})

title {<class 'str'>, <class 'float'>}
content {<class 'str'>, <class 'float'>}


### With each numerical column, how are values distributed?


Numerical columns in `book_df` is `product_id`,	`original_price`,	`current_price`,	`quantity`,	`n_review`,	`avg_rating`,	`pages`.

In [15]:
cols = book_df.select_dtypes([np.number]).columns
min = book_df[cols].min()
missing_percentage = book_df[cols].isnull().sum()/num_rows*100
lower_quartile = book_df[cols].quantile(0.25)
median = book_df[cols].quantile(0.5)
upper_quartile = book_df[cols].quantile(0.75)
max = book_df[cols].max()
nume_col_info_df = pd.concat([missing_percentage,min,lower_quartile,median,upper_quartile,max],axis = 1).T
nume_col_info_df.index = [ "missing_percentage", "min", "lower_quartile", "median", "upper_quartile", "max"]
nume_col_info_df = nume_col_info_df.round(1)
nume_col_info_df

product_id  original_price  current_price  quantity  \
missing_percentage          0.0             0.0            0.0       0.0   
min                    335337.0             0.0            0.0       1.0   
lower_quartile       52785564.0         89000.0        60000.0     173.0   
median              115078649.0        129000.0        88800.0     565.5   
upper_quartile      192908778.0        190000.0       136600.0    2127.8   
max                 207909812.0       1550000.0      1200000.0  671121.0   

                    n_review  avg_rating     pages  
missing_percentage       0.0         0.0       0.2  
min                      0.0         0.0      10.0  
lower_quartile          16.0         4.8     208.0  
median                  74.0         4.8     300.0  
upper_quartile         303.0         5.0     408.0  
max                  11481.0         5.0  192000.0

The maximum row for columns "pages" and "n review" is 192000 and 11481, respectively. Since it's very big, we'll have a look.

In [16]:
book_df[book_df.n_review==11481.0]

product_id                 title                    authors  \
0    74021317  Cây Cam Ngọt Của Tôi  José Mauro de Vasconcelos   

   original_price  current_price  quantity     category  n_review  avg_rating  \
0          108000          64800   53075.0  Tiểu Thuyết     11481         5.0   

   pages              manufacturer  \
0  244.0  Nhà Xuất Bản Hội Nhà Văn   

                                          cover_link  
0  https://salt.tikicdn.com/ts/product/5e/18/24/2...

In [17]:
book_df[book_df.pages==192000.0]

product_id                              title authors  original_price  \
650   201687274  Có Một Nổi Buồn Vừa Ngang Qua Đây      Gã           88000   

     current_price  quantity                         category  n_review  \
650          64900     114.0  Truyện ngắn - Tản văn - Tạp Văn         6   

     avg_rating     pages          manufacturer  \
650         5.0  192000.0  Nhà Xuất Bản Văn Học   

                                            cover_link  
650  https://salt.tikicdn.com/ts/product/08/fd/a5/0...

I checked the webpage, and the information is accurate.

In [18]:
book_df[book_df.original_price==0]

product_id                                           title authors  \
968   203875661  ( Khóa Học ) IELTS Fighter Target 6.5 Speaking     NaN   

     original_price  current_price  quantity            category  n_review  \
968               0              0      56.0  Sách Học Tiếng Anh         0   

     avg_rating  pages           manufacturer  \
968         0.0    NaN  Nhà Xuất Bản Thế Giới   

                                            cover_link  
968  https://salt.tikicdn.com/ts/product/11/b3/ec/4...

There is only one row with a price of 0. We can see that it is a course, so perhaps they are just using it as a means of advertising.

Numerical columns in `comment_df` is `product_id`, `comment_id`, `thank_count`,	`customer_id`,	`rating`.

In [19]:
cols = comment_df.select_dtypes([np.number]).columns
min = comment_df[cols].min()
missing_percentage = comment_df[cols].isnull().sum()/num_rows*100
lower_quartile = comment_df[cols].quantile(0.25)
median = comment_df[cols].quantile(0.5)
upper_quartile = comment_df[cols].quantile(0.75)
max = comment_df[cols].max()
nume_col_info_df = pd.concat([missing_percentage,min,lower_quartile,median,upper_quartile,max],axis = 1).T
nume_col_info_df.index = [ "missing_percentage", "min", "lower_quartile", "median", "upper_quartile", "max"]
nume_col_info_df = nume_col_info_df.round(1)
nume_col_info_df

product_id  comment_id  thank_count  customer_id  rating
missing_percentage          0.0         0.0          0.0          0.0     0.0
min                    335337.0    116197.0          0.0         27.0     1.0
lower_quartile       42459036.0  11576181.0          0.0    5355417.0     5.0
median               77105728.0  15459335.0          0.0   11123876.0     5.0
upper_quartile      147920903.0  17518350.0          0.0   19113314.0     5.0
max                 207852961.0  18392604.0        548.0   28578516.0     5.0

The information is clear and unambiguous.

### With each categorical column, how are values distributed?

Category columns in `book_df` is `category`, `category`, `title`,	`manufacturer`,	`authors`.

In [20]:
cate_cols = set(book_df.columns) - set(book_df.select_dtypes([np.number]).columns)
missing_percentages = np.sum(book_df[cate_cols].isna()) / len(book_df) * 100
num_diff_vals = np.array(book_df[cate_cols].nunique())
diff_vals = np.array(book_df[cate_cols].apply(lambda col: col.dropna().unique()))

cate_cols = pd.DataFrame(data=np.array([missing_percentages, num_diff_vals, diff_vals]),
                            index=['Missing percentage', 'Num diff val', 'Diff val'], columns=cate_cols)
cate_cols

category  \
Missing percentage                                                0.0   
Num diff val                                                      355   
Diff val            [Tiểu Thuyết, Sách tư duy - Kỹ năng sống, Lĩnh...   

                                                           cover_link  \
Missing percentage                                                0.0   
Num diff val                                                     1777   
Diff val            [https://salt.tikicdn.com/ts/product/5e/18/24/...   

                                                         manufacturer  \
Missing percentage                                                0.0   
Num diff val                                                       42   
Diff val            [Nhà Xuất Bản Hội Nhà Văn, Nhà Xuất Bản Thế Gi...   

                                                              authors  \
Missing percentage                                           7.878447   
Num diff val                                                     1083   
Diff val            [José Mauro de Vasconcelos, Nguyễn Đoàn Minh T...   

                                                                title  
Missing percentage                                                0.0  
Num diff val                                                     1769  
Diff val            [Cây Cam Ngọt Của Tôi, Hành Tinh Của Một Kẻ Ng...

Category columns in `comment_df` is `title`	, `content`.

In [21]:
cate_cols = set(comment_df.columns) - set(comment_df.select_dtypes([np.number]).columns)
missing_percentages = np.sum(comment_df[cate_cols].isna()) / len(comment_df) * 100
num_diff_vals = np.array(comment_df[cate_cols].nunique())
diff_vals = np.array(comment_df[cate_cols].apply(lambda col: col.dropna().unique()))

cate_cols = pd.DataFrame(data=np.array([missing_percentages, num_diff_vals, diff_vals]),
                            index=['Missing percentage', 'Num diff val', 'Diff val'], columns=cate_cols)
cate_cols

content  \
Missing percentage                                                0.0   
Num diff val                                                    75681   
Diff val            [Có những người bước đến, họ lấp đầy hạnh phúc...   

                                                                title  
Missing percentage                                                0.0  
Num diff val                                                     4164  
Diff val            [Cực kì hài lòng, Bình thường, Một cuốn sách đ...

Lets look at `manufacturer` columns. 

In [22]:
book_df.manufacturer.value_counts()

nan                                              265
Nhà Xuất Bản Thế Giới                            249
NXB Trẻ                                          145
Nhà Xuất Bản Hà Nội                              128
Nhà Xuất Bản Dân Trí                             126
Nhà Xuất Bản Kim Đồng                            113
Nhà Xuất Bản Hội Nhà Văn                         111
Nhà Xuất Bản Thanh Niên                           98
Nhà Xuất Bản Tổng hợp TP.HCM                      92
Nhà Xuất Bản Lao Động                             90
Nhà Xuất Bản Văn Học                              77
Nhà Xuất Bản Hồng Đức                             57
Nhà Xuất Bản Công Thương                          46
Nhà Xuất Bản Phụ Nữ Việt Nam                      38
Nhà Xuất Bản Phụ Nữ                               30
Nhà Xuất Bản Đà Nẵng                              17
Nhà Xuất Bản Đại Học Quốc Gia Hà Nội              13
Nhà Xuất Bản Mỹ Thuật                              8
Nhà Xuất Bản Lao Động Xã Hội                  

We will replace `NXB` by `Nhà Xuất Bản`, `hong duc` by `Nhà Xuất Bản Hồng Đức` and `ĐHQG`:`Đại Học Quốc Gia`

In [23]:
book_df['manufacturer'] = book_df['manufacturer'].replace({'NXB':'Nhà Xuất Bản','hong duc':'Nhà Xuất Bản Hồng Đức','ĐHQG':'Đại Học Quốc Gia'}, regex=True)

I checked the website and found that the authors of these books are listed as "."

In [24]:
book_df[book_df['authors']=='.'][:5]

product_id                                              title authors  \
145     73125816  Digital Marketing - Từ Chiến Lược Đến Thực Thi...       .   
270     54614797                       Có Một Ngày, Bố Mẹ Sẽ Già Đi       .   
883     40774181  Combo Sách Tiếng hàn tổng hợp dành cho người V...       .   
1194   145977554   Combo 2 Cuốn: Thần Số Học + Thần Số Học Ứng Dụng       .   
1239    45349383  Hashtag No.2 Fashion - Kinh Doanh Thời Trang T...       .   

      original_price  current_price  quantity  \
145           220000         154000    4487.0   
270            96000          58520   10530.0   
883           220000         147400    5295.0   
1194          348000         244000     403.0   
1239          200000         140000    1768.0   

                             category  n_review  avg_rating  pages  \
145        Sách Marketing - Bán hàng        865         4.8  250.0   
270   Truyện ngắn - Tản văn - Tạp Văn      2252         4.8  296.0   
883                Sách Học Tiếng Hàn      1305         5.0  436.0   
1194      Sách Chiêm Tinh - Horoscope        62         4.8    NaN   
1239       Sách Marketing - Bán hàng        213         4.8  132.0   

                 manufacturer  \
145     Nhà Xuất Bản Lao Động   
270     Nhà Xuất Bản Thế Giới   
883                       nan   
1194  Nhà Xuất Bản Thanh Niên   
1239     Nhà Xuất Bản Dân Trí   

                                             cover_link  
145   https://salt.tikicdn.com/ts/product/d0/b7/ac/8...  
270   https://salt.tikicdn.com/ts/product/89/f1/fc/1...  
883   https://salt.tikicdn.com/ts/product/6d/97/55/e...  
1194  https://salt.tikicdn.com/ts/product/e6/21/58/b...  
1239  https://salt.tikicdn.com/ts/product/a0/fe/92/7...

Let's look at `category` field.

In [25]:
book_df['category'].value_counts()

Sách tư duy - Kỹ năng sống              295
Tiểu Thuyết                             133
Truyện ngắn - Tản văn - Tạp Văn         109
Sách nghệ thuật sống đẹp                 58
Sách kỹ năng làm việc                    56
                                       ... 
Birdmen Tập 13 [Tặng Kèm Postcard]        1
Bên Rặng Tuyết Sơn (Tái Bản)              1
Ngược Dòng Tiền Kiếp                      1
One Piece - Tập 100 :  Haki Bá vương      1
Bá Tước Tiểu Thư - Tập 5                  1
Name: category, Length: 355, dtype: int64

The 'category' columns in this dataset may have been taken incorrectly by the author, resulting in some incorrect categories. We looked over the website and gathered data on every category a book might fall under.There are many categories and subcategories, however practically every book in our dataset is "Sách tiếng việt," thus we only focus on its subcategories.

In [26]:
cate = {'Sách văn học': ['Du ký','Light novel','Phê Bình - Lý Luận Văn Học', 'Phóng sự - Ký sự - Bút ký', 'Tác phẩm kinh điển','Thơ',
 'Tiểu sử - Hồi ký', 'Tiểu Thuyết','Tranh Truyện',
  'Truyện cổ tích - Ngụ ngôn','Truyện cười', 'Truyện dài', 'Truyện đam mỹ',
  'Truyện Giả tưởng - Huyền Bí - Phiêu Lưu', 'Truyện kiếm hiệp',
  'Truyện kinh dị','Truyện ngắn - Tản văn - Tạp Văn',
  'Truyện ngôn tình', 'Truyện trinh thám', 'Ca dao - Tục ngữ', 'Sách ảnh'],
 'Sách kinh tế': ['Bài học kinh doanh',
  'Sách doanh nhân',
  'Sách khởi nghiệp',
  'Sách kinh tế học',
  'Sách kỹ năng làm việc',
  'Sách Marketing - Bán hàng',
  'Sách quản trị, lãnh đạo',
  'Sách quản trị nhân lực',
  'Sách tài chính, kế toán',
  'Sách tài chính, tiền tệ'],
 'Sách thiếu nhi': ['Đạo đức - Kỹ năng sống',
  'Kiến thức - Bách khoa',
  'Tô màu - Luyện chữ',
  'Truyện cổ tích',
  'Truyện kể cho bé',
  'Truyện tranh Ehon',
  'Truyện tranh thiếu nhi',
  'Văn học thiếu nhi'],
 'Sách kỹ năng sống': ['Sách tư duy - Kỹ năng sống',
  'Sách nghệ thuật sống đẹp',
  'Sách hướng nghiệp - Kỹ năng mềm'],
 'Sách Bà mẹ - Em bé': ['Sách Thai Giáo',
  'Sách Kiến Thức - Kỹ Năng Cho Trẻ',
  'Sách Làm Cha Mẹ',
  'Sách Dinh Dưỡng - Sức Khỏe Cho Bé'],
 'Sách Giáo Khoa - Giáo Trình': ['Sách Giáo Khoa Cấp 1',
  'Sách Giáo Khoa Cấp 2',
  'Sách Giáo Khoa Cấp 3',
  'Giáo Trình Đại Học - Cao Đẳng'],
 'Sách Học Ngoại Ngữ': ['Sách Học Tiếng Anh',
  'Sách Học Tiếng Nhật',
  'Sách Học Tiếng Hàn',
  'Sách Học Tiếng Hoa',
  'Ngoại Ngữ Khác'],
 'Sách Tham Khảo': ['Sách tham khảo cấp I',
  'Sách tham khảo cấp II',
  'Sách tham khảo cấp III',
  'Sách Luyện Thi Đại Học - Cao Đẳng',
  'Thẻ luyện thi',
  'Sách chuyên đề'],
 'Từ Điển': ['Từ Điển Tiếng Anh',
  'Từ Điển Tiếng Trung',
  'Từ Điển Tiếng Hàn',
  'Từ Điển Tiếng Nhật',
  'Từ Điển Tiếng Pháp',
  'Từ Điển Tiếng Việt',
  'Từ Điển Ngôn Ngữ Khác'],
 'Sách Kiến Thức Tổng Hợp': ['Kiến Thức Bách Khoa',
  'Lĩnh vực khác',
  'Sách Chiêm Tinh - Horoscope',
  'Sách giáo dục',
  'Sách Phong Thủy - Kinh Dịch',
  'Triết Học'],
 'Sách Khoa Học - Kỹ Thuật': ['Sách Khoa Học - Kỹ Thuật'],
 'Sách Lịch sử': ['Lịch Sử Việt Nam', 'Lịch Sử Thế Giới'],
 'Điện Ảnh - Nhạc - Họa': ['Âm Nhạc - Điện Ảnh',
  'Mỹ Thuật - Kiến Trúc',
  'Sách Tô Màu Dành Cho Người Lớn'],
 'Truyện Tranh, Manga, Comic': ['Truyện Tranh, Manga, Comic'],
 'Sách Tôn Giáo - Tâm Linh': ['Sách Tôn Giáo - Tâm Linh'],
 'Sách Văn Hóa - Địa Lý - Du Lịch': ['Sách Danh Nhân',
  'Sách Địa Danh - Du Lịch',
  'Sách Phong Tục - Tập Quán'],
 'Sách Chính Trị - Pháp Lý': ['Luật - Văn Bản Luật', 'Lý Luận Chính Trị'],
 'Sách Nông - Lâm - Ngư Nghiệp': ['Sách Nông - Lâm - Ngư Nghiệp'],
 'Sách Công Nghệ Thông Tin': ['Tin Học Văn Phòng',
  'Lập Trình',
  'Thiết Kế - Đồ Họa'],
 'Thể Dục - Thể Thao': ['Thể Dục - Thể Thao'],
 'Sách Y Học': ['Sách Y Học'],
 'Tạp Chí - Catalogue': ['Tạp Chí - Catalogue'],
 'Sách Tâm lý - Giới tính': ['Sách Tâm Lý Tuổi Teen',
  'Sách Hôn Nhân - Giới Tính'],
 'Sách Thường Thức - Gia Đình': ['Chăm sóc sức khỏe',
  'May - Thêu - Thời Trang',
  'Sách Làm Đẹp',
  'Sách Nấu ăn']}

In [27]:

category = pd.read_csv('Category')
category.columns = ['Category','SubCategory']
category['SubCategory'] = category['SubCategory'].apply(lambda x: x.split(';'))

We will gather those books once more, but only in the category field, to fix the inaccurate values.

In [28]:
cate_list = category['SubCategory'].explode('SubCategory').to_list()
index = book_df[~book_df.category.isin(cate_list)].product_id.to_list()

In [29]:
repair_cate = pd.read_csv('RepairCategory')
repair_cate.index = repair_cate.product_id
repair_cate.head()

Unnamed: 0  Unnamed: 0.1  product_id  \
product_id                                         
69219919             0            23    69219919   
95470626             1            61    95470626   
188695904            2            73   188695904   
203957925            3            94   203957925   
207475540            4            97   207475540   

                                                        title  \
product_id                                                      
69219919                        Muôn Kiếp Nhân Sinh (Khổ Nhỏ)   
95470626    Combo 2 Cuốn: Muôn Kiếp Nhân Sinh + Muôn Kiếp ...   
188695904                                        GIẬN (TBL22)   
203957925                                  Monster #8 - Tập 6   
207475540             Blue Flag Tập 8 [Tặng Kèm Bookmark PVC]   

                    authors  original_price  current_price  quantity  \
product_id                                                             
69219919       Nguyên Phong           98000          72000   12009.0   
95470626       Nguyên Phong          436000         248500    5947.0   
188695904   Thích Nhất Hạnh          135000         116000    1802.0   
203957925   Naoya Matsumoto           58000          49200    1438.0   
207475540             Kaito           45000          45000     143.0   

                              category  n_review  avg_rating  pages  \
product_id                                                            
69219919      Sách Tôn Giáo - Tâm Linh      2603         4.8  482.0   
95470626      Sách Tôn Giáo - Tâm Linh      1207         4.8  960.0   
188695904     Sách Tôn Giáo - Tâm Linh       145         5.0  248.0   
203957925   Truyện Tranh, Manga, Comic       132         5.0  188.0   
207475540   Truyện Tranh, Manga, Comic        15         4.7  244.0   

                            manufacturer  \
product_id                                 
69219919    Nhà Xuất Bản Tổng hợp TP.HCM   
95470626    Nhà Xuất Bản Tổng hợp TP.HCM   
188695904          Nhà Xuất Bản Thế Giới   
203957925          Nhà Xuất Bản Hồng Đức   
207475540          Nhà Xuất Bản Kim Đồng   

                                                   cover_link  
product_id                                                     
69219919    https://salt.tikicdn.com/media/catalog/product...  
95470626    https://salt.tikicdn.com/media/catalog/product...  
188695904   https://salt.tikicdn.com/ts/product/56/4a/55/d...  
203957925   https://salt.tikicdn.com/ts/product/78/98/79/d...  
207475540   https://salt.tikicdn.com/ts/product/c5/0d/50/5...

In [30]:
book_df = book_df.set_index('product_id')
book_df.loc[index,'category']  = repair_cate.loc[index,'category'].to_list()
book_df.loc[index,'category'] 

product_id
69219919       Sách Tôn Giáo - Tâm Linh
95470626       Sách Tôn Giáo - Tâm Linh
188695904      Sách Tôn Giáo - Tâm Linh
203957925    Truyện Tranh, Manga, Comic
207475540    Truyện Tranh, Manga, Comic
                        ...            
138504802     Sách Marketing - Bán hàng
995310                       Sách Y Học
190464440    Truyện Tranh, Manga, Comic
200619499    Truyện Tranh, Manga, Comic
74684484      Sách Marketing - Bán hàng
Name: category, Length: 334, dtype: object

We will check them again, and there are 2 rows where the category is still missing from our list.

In [31]:
 book_df[~book_df.category.isin(cate_list)]

title        authors  \
product_id                                                                     
161409461   Thùng 48 hộp sữa tươi tiệt trùng có đường TH T...            NaN   
204374127   Tạp chí Forbes Việt Nam - Số 111 (Tháng 11.202...  nhieu tac gia   

            original_price  current_price  quantity              category  \
product_id                                                                  
161409461           406800         386000    1372.0  Sữa tiệt trùng (UHT)   
204374127            95000          95000      38.0       Sách tiếng Việt   

            n_review  avg_rating  pages     manufacturer  \
product_id                                                 
161409461        117         5.0    NaN              nan   
204374127          4         5.0  140.0  Forbes Việt Nam   

                                                   cover_link  
product_id                                                     
161409461   https://salt.tikicdn.com/ts/product/66/33/87/5...  
204374127   https://salt.tikicdn.com/ts/product/03/f0/8d/4...

We can see that `Thùng 48 hộp sữa tươi tiệt trùng có đường TH` is not a book so we just remove it from our dataset 

In [32]:
book_df = book_df.drop(161409461)

In [33]:
# join = comment_df.merge(book_df,left_on='product_id',right_on='product_id')
# book_df = book_df.reset_index(drop=True)
comment_df.product_id.value_counts()[comment_df.product_id.value_counts()>5]
# comment_df.rating.value_counts()
# comment_df

107802183    91
50773990     90
163165281    90
197587849    90
11189120     90
             ..
201307595     6
201580268     6
35283865      6
595703        6
204363219     6
Name: product_id, Length: 1828, dtype: int64

In [34]:
data = comment_df.merge(book_df,left_on='product_id',right_on='product_id',how='inner')
pivot_table = data.pivot_table(index = ["title_y"],columns = ["customer_id"],values = "rating").fillna(0)
pivot_table.head(10)

customer_id                                         27        120       \
title_y                                                                  
 Cứ Làm Đi!                                              0.0       0.0   
 Để Trở Thành Người Bán Hàng Giỏi Nhất Thế Giới          0.0       0.0   
"999 Lá Thư Gửi Cho Chính Mình” - Phiên Bản Mini         0.0       0.0   
"Cậu" Ma Nhà Xí Hanako - Tập 0                           0.0       0.0   
"Đức Phật" nơi công sở                                   0.0       0.0   
#Hashtag 01: Drink - Kinh Doanh Đồ Uống Tại Thị...       0.0       0.0   
(Bìa cứng) HIMALAYA NHỮNG CUỘC PHIÊU LƯU - THIỀ...       0.0       0.0   
10 vạn câu hỏi vì sao - trọn bộ 5 tập                    0.0       0.0   
100+ Bài Tập Kích Thích Não Bộ Hiệu Quả  40 Ngà...       0.0       0.0   
101 Lời Khuyên Tài Chính Cá Nhân Từ Thái Phạm            0.0       0.0   

customer_id                                         247       285       \
title_y                                                                  
 Cứ Làm Đi!                                              0.0       0.0   
 Để Trở Thành Người Bán Hàng Giỏi Nhất Thế Giới          0.0       0.0   
"999 Lá Thư Gửi Cho Chính Mình” - Phiên Bản Mini         0.0       0.0   
"Cậu" Ma Nhà Xí Hanako - Tập 0                           0.0       0.0   
"Đức Phật" nơi công sở                                   0.0       0.0   
#Hashtag 01: Drink - Kinh Doanh Đồ Uống Tại Thị...       0.0       0.0   
(Bìa cứng) HIMALAYA NHỮNG CUỘC PHIÊU LƯU - THIỀ...       0.0       0.0   
10 vạn câu hỏi vì sao - trọn bộ 5 tập                    0.0       0.0   
100+ Bài Tập Kích Thích Não Bộ Hiệu Quả  40 Ngà...       0.0       0.0   
101 Lời Khuyên Tài Chính Cá Nhân Từ Thái Phạm            0.0       0.0   

customer_id                                         393       406       \
title_y                                                                  
 Cứ Làm Đi!                                              0.0       0.0   
 Để Trở Thành Người Bán Hàng Giỏi Nhất Thế Giới          0.0       0.0   
"999 Lá Thư Gửi Cho Chính Mình” - Phiên Bản Mini         0.0       0.0   
"Cậu" Ma Nhà Xí Hanako - Tập 0                           0.0       0.0   
"Đức Phật" nơi công sở                                   0.0       0.0   
#Hashtag 01: Drink - Kinh Doanh Đồ Uống Tại Thị...       0.0       0.0   
(Bìa cứng) HIMALAYA NHỮNG CUỘC PHIÊU LƯU - THIỀ...       0.0       0.0   
10 vạn câu hỏi vì sao - trọn bộ 5 tập                    0.0       0.0   
100+ Bài Tập Kích Thích Não Bộ Hiệu Quả  40 Ngà...       0.0       0.0   
101 Lời Khuyên Tài Chính Cá Nhân Từ Thái Phạm            0.0       0.0   

customer_id                                         479       544       \
title_y                                                                  
 Cứ Làm Đi!                                              0.0       0.0   
 Để Trở Thành Người Bán Hàng Giỏi Nhất Thế Giới          0.0       0.0   
"999 Lá Thư Gửi Cho Chính Mình” - Phiên Bản Mini         0.0       0.0   
"Cậu" Ma Nhà Xí Hanako - Tập 0                           0.0       0.0   
"Đức Phật" nơi công sở                                   0.0       0.0   
#Hashtag 01: Drink - Kinh Doanh Đồ Uống Tại Thị...       0.0       0.0   
(Bìa cứng) HIMALAYA NHỮNG CUỘC PHIÊU LƯU - THIỀ...       0.0       0.0   
10 vạn câu hỏi vì sao - trọn bộ 5 tập                    0.0       0.0   
100+ Bài Tập Kích Thích Não Bộ Hiệu Quả  40 Ngà...       0.0       0.0   
101 Lời Khuyên Tài Chính Cá Nhân Từ Thái Phạm            0.0       0.0   

customer_id                                         622       661       ...  \
title_y                                                                 ...   
 Cứ Làm Đi!                                              0.0       0.0  ...   
 Để Trở Thành Người Bán Hàng Giỏi Nhất Thế Giới          0.0       0.0  ...   
"999 Lá Thư Gửi Cho Chính Mình” - Phiên Bản Mini         0.0       0.0  ...   
"Cậu" Ma Nhà Xí Hanako - Tập 0                   

In [40]:
book_df

title  \
product_id                                                      
74021317                                 Cây Cam Ngọt Của Tôi   
184466860                     Hành Tinh Của Một Kẻ Nghĩ Nhiều   
73787185                             Những Tù Nhân Của Địa Lý   
52789367                           Nhà Giả Kim (Tái Bản 2020)   
147920903                    Một Thoáng Ta Rực Rỡ Ở Nhân Gian   
...                                                       ...   
192949122   7 Thói Quen Hiệu Quả - The 7 Habits Of Highly ...   
157579054   Các Vấn Đề Của Triết Học (The Problems Of Phil...   
20704715    Để Con Được Chích - Hiểu Hết Về Vắc Xin Và Miễ...   
131144753                       Nhà Lãnh Đạo 360 Độ (Tái Bản)   
58971672    The Best of Chicken Soup - Tuyển Tập Những Câu...   

                              authors  original_price  current_price  \
product_id                                                             
74021317    José Mauro de Vasconcelos          108000          64800   
184466860        Nguyễn Đoàn Minh Thư           86000          59900   
73787185                 Tim Marshall          210000         126000   
52789367                 Paulo Coelho           79000          47400   
147920903                         NaN          135000          81000   
...                               ...             ...            ...   
192949122             Stephen R Covey          250000         191200   
157579054            Bertrand Russell           85000          48900   
20704715                     Uyên Bùi          108000          43200   
131144753             John C. Maxwell          179000         107500   
58971672                Jack Canfield          138000          91000   

            quantity                    category  n_review  avg_rating  pages  \
product_id                                                                      
74021317     53075.0                 Tiểu Thuyết     11481         5.0  244.0   
184466860     7929.0  Sách tư duy - Kỹ năng sống       780         4.8  184.0   
73787185     17896.0               Lĩnh vực khác      3623         4.8  430.0   
52789367     24668.0          Tác phẩm kinh điển      5131         5.0  228.0   
147920903    10000.0                 Tiểu Thuyết      1636         4.8  304.0   
...              ...                         ...       ...         ...    ...   
192949122       76.0  Sách tư duy - Kỹ năng sống         7         5.0    NaN   
157579054      147.0                   Triết Học        20         5.0  236.0   
20704715       931.0             Sách Làm Cha Mẹ        47         4.6  296.0   
131144753      378.0     Sách quản trị, lãnh đạo        33         5.0  436.0   
58971672       954.0    Sách nghệ thuật sống đẹp       176         5.0  256.0   

                            manufacturer  \
product_id                                 
74021317        Nhà Xuất Bản Hội Nhà Văn   
184466860          Nhà Xuất Bản Thế Giới   
73787185        Nhà Xuất Bản Hội Nhà Văn   
52789367             Nhà Xuất Bản Hà Nội   
147920903       Nhà Xuất Bản Hội Nhà Văn   
...                                  ...   
192949122   Nhà Xuất Bản Tổng hợp TP.HCM   
157579054   Nhà Xuất Bản Khoa Học Xã Hội   
20704715                             nan   
131144753          Nhà Xuất Bản Lao Động   
58971672    Nhà Xuất Bản Tổng hợp TP.HCM   

                                                   cover_link  
product_id                                                     
74021317    https://salt.tikicdn.com/ts/product/5e/18/24/2...  
184466860   https://salt.tikicdn.com/ts/product/d7/99/24/1...  
73787185    https://salt.tikicdn.com/ts/product/8d/96/9e/c...  
52789367    https://salt.tikicdn.com/ts/product/45/3b/fc/a...  
147920903   https://salt.tikicdn.com/ts/product/2f/b5/4e/a...  
...                                                       ...  
192949122   https://salt.tikicdn.com/ts/product/ba/c3/c0/1...  
157579054   https://salt.tikicdn.com/ts/product/a5/52/7f/3...  
207047